daemon/graphdriver/copy/copy.go

```diff
@@ -107,156 +120,393 @@ func copyXattr(srcPath, dstPath, attr string) error {
+			return err
+		}
 
-	dirsToSetMtimes := list.New()
-	err := filepath.Walk(srcDir, func(srcPath string, f os.FileInfo, err error) error {
+	case os.ModeSymlink:
+		link, err := os.Readlink(srcPath)
 		if err != nil {
 			return err
 		}
 
-		// Rebase path
-		relPath, err := filepath.Rel(srcDir, srcPath)
-		if err != nil {
+		if err := os.Symlink(link, dstPath); err != nil {
 			return err
 		}
 
-		dstPath := filepath.Join(dstDir, relPath)
-		if err != nil {
+	case os.ModeNamedPipe:
+		fallthrough
+	case os.ModeSocket:
+		if err := unix.Mkfifo(dstPath, stat.Mode); err != nil {
 			return err
 		}
 
-		stat, ok := f.Sys().(*syscall.Stat_t)
-		if !ok {
-			return fmt.Errorf("Unable to get raw syscall.Stat_t data for %s", srcPath)
+	case os.ModeDevice:
+		if rsystem.RunningInUserNS() {
+			// cannot create a device if running in user namespace
+			return nil
+		}
+		if err := unix.Mknod(dstPath, stat.Mode, int(stat.Rdev)); err != nil {
+			return err
 		}
 
-		isHardlink := false
+	default:
+		return fmt.Errorf("unknown file type for %s", srcPath)
+	}
+
+	return copyMetadata(dstPath, srcPath, f, stat, copyXattrs)
+}
+
+func fileCopyWorker(ctx context.Context, fileCopyInfoChan chan *fileCopyInfo, copyMode Mode, copyXattrs bool) error {
+	copyWithFileRange := true
+	copyWithFileClone := true
 
-		switch f.Mode() & os.ModeType {
-		case 0: // Regular file
-			id := fileID{dev: stat.Dev, ino: stat.Ino}
-			if copyMode == Hardlink {
-				isHardlink = true
-				if err2 := os.Link(srcPath, dstPath); err2 != nil {
-					return err2
+	// Just a dumb wrapper around the above function which does the heavy lifting
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case fci, ok := <-fileCopyInfoChan:
+			if !ok {
+				return nil
+			}
+			switch fci.mode {
+			case dirMode:
+				if err := copyMetadataShared(fci.dstPath, fci.srcPath, fci.fileInfo, fci.stat, copyXattrs); err != nil {
+					return err
 				}
-			} else if hardLinkDstPath, ok := copiedFiles[id]; ok {
-				if err2 := os.Link(hardLinkDstPath, dstPath); err2 != nil {
-					return err2
+			case fileMode:
+				if err := doFileCopy(fci.srcPath, fci.dstPath, fci.fileInfo, fci.stat, fci.sharedInode, &copyWithFileRange, &copyWithFileClone, copyXattrs, copyMode); err != nil {
+					return err
 				}
-			} else {
-				if err2 := copyRegular(srcPath, dstPath, f, &copyWithFileRange, &copyWithFileClone); err2 != nil {
-					return err2
+			case updateDirTimeMode:
+				ts := []syscall.Timespec{fci.stat.Atim, fci.stat.Mtim}
+				if err := system.LUtimesNano(fci.dstPath, ts); err != nil {
+					return err
 				}
-				copiedFiles[id] = dstPath
 			}
+		}
 
-		case os.ModeDir:
-			if err := os.Mkdir(dstPath, f.Mode()); err != nil && !os.IsExist(err) {
-				return err
-			}
+	}
+}
 
-		case os.ModeSymlink:
-			link, err := os.Readlink(srcPath)
-			if err != nil {
-				return err
-			}
+// readDirNames reads the directory named by dirname and returns
+// a sorted list of directory entries.
```

Comment:
Is this accurate? I don't see where it's sorted (nor does Golang's description mention sorting); https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/os/dir.go#L29-L40

Not a big issue, just double-checking so that we don't document behavior that isn't guaranteed / part of the contract